File: sdks/js/langgraph/cases/10-binary-content-redaction.js
/**
 * 10-binary-content-redaction: Binary Content Redaction Test
 *
 * Tests that when binary data (such as images) is sent to an LLM via LangGraph,
 * Sentry correctly redacts the binary content in the captured span data and
 * replaces it with a substitute marker ("[Blob substitute]").
 */

const { Sentry } = require("../setup");
const { ChatOpenAI } = require("@langchain/openai");
const { createReactAgent } = require("@langchain/langgraph/prebuilt");
const { HumanMessage } = require("@langchain/core/messages");
const { runTestCase } = require("../../_test-utils/test-runner.cjs");

/**
 * Creates a minimal valid PNG image (10x10 red square).
 * Returns base64-encoded PNG data.
 */
function createMinimalPng() {
  // Minimal 10x10 red PNG image (base64)
  // This is a pre-generated minimal valid PNG to avoid needing canvas/sharp dependencies
  const minimalPngBase64 =
    "iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAFUlEQVR42mP8z8BQz0AEYBxVSF+FABJADveWkH6oAAAAAElFTkSuQmCC";
  
  return minimalPngBase64;
}

async function testLogic(inputs) {
  const { model, image_type } = inputs;

  // Create LLM instance
  const llm = new ChatOpenAI({
    modelName: model,
    apiKey: process.env.OPENAI_API_KEY,
  });

  // Create a simple react agent with no tools
  const agent = createReactAgent({ llm, tools: [] });

  // Create binary image data
  const base64Image = createMinimalPng();

  // Create message with image content using LangChain's multimodal format
  const message = new HumanMessage({
    content: [
      {
        type: "image_url",
        image_url: { url: `data:image/${image_type};base64,${base64Image}` },
      },
      {
        type: "text",
        text: "What color is this image? Answer in one word.",
      },
    ],
  });

  // Invoke the agent with image message
  const result = await agent.invoke({
    messages: [message],
  });

  // Extract the AI's response from the result
  const resultMessages = result.messages;
  const lastMessage = resultMessages[resultMessages.length - 1];
  const text = lastMessage.content;

  if (!text) {
    throw new Error("No completion returned from LangGraph");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    Response: ${text}`);
    console.log(
      `    Sent image with ${base64Image.length} characters of base64 data`
    );
  }
}

module.exports = runTestCase("10-binary-content-redaction", testLogic, Sentry);

File: sdks/py/langgraph/cases/10-binary-content-redaction.py
"""
10-binary-content-redaction: Binary Content Redaction Test

Tests that when binary data (such as images) is sent to an LLM via LangGraph,
Sentry correctly redacts the binary content in the captured span data and
replaces it with a substitute marker.
"""

import os
import base64
from pathlib import Path
from langgraph.prebuilt import create_react_agent
from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage
from test_runner import run_test_case


async def test_logic(inputs):
    """The actual test logic"""
    model = inputs["model"]

    llm = ChatOpenAI(model=model, api_key=os.getenv("OPENAI_API_KEY"))

    # Create a simple react agent with no tools
    agent = create_react_agent(llm, tools=[])

    # Load static test image
    # Path: cases -> langgraph -> py -> sdks -> repo_root
    repo_root = Path(__file__).parent.parent.parent.parent.parent
    image_path = repo_root / "shared" / "test-assets" / "test-image-10x10-red.png"

    with open(image_path, "rb") as f:
        image_data = f.read()
    base64_image = base64.standard_b64encode(image_data).decode("utf-8")

    # Create message with image content
    message = HumanMessage(
        content=[
            {
                "type": "image_url",
                "image_url": {"url": f"data:image/png;base64,{base64_image}"},
            },
            {
                "type": "text",
                "text": "What color is this image? Answer in one word.",
            },
        ]
    )

    result = agent.invoke({"messages": [message]})

    # Extract the AI's response from the result
    text = result["messages"][-1].content

    if not text:
        raise Exception("No output returned from LangGraph")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    Response: {text}")
        print(f"    Sent image with {len(image_data)} bytes of binary data")

    return text


# Export test case functions
test_case = run_test_case("10-binary-content-redaction", test_logic)
main = test_case["main"]
assert_sentry = test_case["assert_sentry"]
